### PR TITLE
[FrameworkBundle] added a protected shortcut getParameter() method in the base Controller class.

### PR DESCRIPTION
**Pull request by hhamon**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | ~ |

### CODE REVIEW

**Comment by hhamon** [comment]
I don't know how this will be welcome but I always find annoying to have to explicitly access the `container` property to get a global parameter from an action.

I'm also wondering of naming the method to `getConfigParameter()` or `getGlobalParameter()`.

**Comment by javiereguiluz** [comment]
@hhamon :+1: this has bugged me a lot of times!

In case this proposal is accepted, my vote would be for the shortest `->getParameter()` method name.

File: src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php [hunk 1]
```diff
@@ -345,11 +345,23 @@ public function get($id)
         if ('request' === $id) {
             trigger_error('The "request" service is deprecated and will be removed in 3.0. Add a typehint for Symfony\\Component\\HttpFoundation\\Request to your controller parameters to retrieve the request instead.', E_USER_DEPRECATED);
         }
-        
+
         return $this->container->get($id);
     }
 
     /**
+     * Gets a global configuration parameter.
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php]: Maybe we could tweak this description to replace `global` by `container`.

**Comment by ogizanagi** [comment]
:+1: for the `getParameter` method name.

**Comment by aitboudad** [comment]
what do you think about adding ``` hasParameter ``` too ?

**Comment by javiereguiluz** [comment]
@aitboudad good point! Maybe we're complicating things a lot but, what if we mimick the ParamaterBag behavior and change this `getParameter(string $name)` by `getParameter(string $name, mixed $default = null)`?

File: src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php (hunk 1 above)
Comment (hhamon) [inline src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php]: Good idea :)

**Comment by hhamon** [comment]
I'm +1 for the default value as the second argument. Makes more sense in the Controller layer to avoid cascading `if` statements.

**Comment by phansys** [comment]
IMHO, ```getParameter``` can be quickly misunderstood since it is in a controller action that process an HTTP request. I mean, the first thought would be to receive a request parameter.

**Comment by hhamon** [comment]
Good point! This is the reason why I wonder about the naming of the method...

**Comment by ogizanagi** [comment]
`$this->getContainerParameter($name)` then ?

I'm not sure it's better than `$this->container->getParameter($name)`... unless it adds the `ParameterBag` mimic.

> IMHO, getParameter can be quickly misunderstood since it is in a controller action that process an HTTP request. I mean, the first thought would be to receive a request parameter.

Reading at the PR title, I thought about the request parameters...
But then, the same thought could be applied to the `get` and `has` methods, which can be misunderstood for the same reasons and might suggest retrieving query parameters...

**Comment by aitboudad** [comment]
``` getConfigParameter ``` ?

**Comment by javiereguiluz** [comment]
I don't really see any big issue with the `getParameter()` name. Even newcomers know that when you use  `$this->...` inside a controller, you never refer to the request but the application. People usually don't have any problem understanding the `$request $request` typehinting mechinsm of obtaining the request.

**Comment by phansys** [comment]
@aitboudad , ```getConfigParameter()``` seems more consistent to me. @javiereguiluz, I don't believe this is actually an issue. I only share my opinion about the method naming inside the controller scope. BTW, despite as how it will called, I think it's a nice addition.

**Comment by hhamon** [comment]
Why not naming it `getConfiguration()` to make it clearer?

**Comment by hhamon** [comment]
@weaverryan any thoughts about naming things on this PR please?

**Comment by xabbuh** [comment]
I wouldn't use `getConfiguration()` as the method's name. It could easily be confused with what you add in the application configuration files. To me, `getParameter()` is just fine.

**Comment by jakzal** [comment]
:+1: for `getParameter()`. This is actually what people intuitively try to use on the basic Symfony training course, once they're shown the `get()` shortcut.

However, I'm concerned with introduction of ParameterBag-like default value for two reasons. Firstly, this behaviour would be inconsistent with the way container works. It'll lead to WTF moments when accessing container parameters directly. Secondly, I think in this case it's better to fail early and throw an exception to show the application is misconfigured.

**Comment by javiereguiluz** [comment]
After reading @jakzal comments I realize how wrong was my proposal about providing a default mechanism for container parameters. In the case of request params makes sense in a lot of use cases (e.g. `->query->get('page', 1)`) but in the case of the container, if a required parameter doesn't exist, the application should fail.

**Comment by hhamon** [comment]
Hi Jakub,

Thanks for your reply. So you mean writing something like the following:

```
function getParameter($name)
{
    if (!$this->container->hasParameter($name)) {
        throw new \InvalidArgumentException(sprintf('Invalid parameter "%s".', $name)); 
    }

    return $this->container->getParameter($name);
}
```

**Comment by jakzal** [comment]
Container's `ParameterBag` already [throws an exception if you try to access a parameter that doesn't exist](https://github.com/symfony/DependencyInjection/blob/master/ParameterBag/ParameterBag.php#L93-L107). This could be a simple proxy method:

```php
public function getParameter($name)
{
    return $this->container->getParameter($name);
}
```

**Comment by hhamon** [comment]
This is already what I've done in the PR with a protected proxy method.

**Comment by jakzal** [comment]
Well, no. Currently the method checks if parameter exists and uses a default value otherwise.

**Comment by hhamon** [comment]
Oh yes! Since I've updated the code according to @javiereguiluz proposal. I'm reverting it.

**Comment by hhamon** [comment]
@jakzal fixed thanks.

**Comment by xabbuh** [comment]
:+1:

**Comment by jakzal** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @hhamon.

**Pull request merged**